Algorithm.Framework/Portfolio/EqualWeightingPortfolioConstructionModel.cs

```diff
@@ -40,50 +53,55 @@ public class EqualWeightingPortfolioConstructionModel : PortfolioConstructionMod
 
-            // Get last insight that haven't expired of each symbol that is still in the universe
-            var activeInsights = (from insight in _insightCollection
-                                  // Remove expired insights
-                                  where insight.CloseTimeUtc > algorithm.UtcTime
-                                  // Force one group per symbol
-                                  group insight by insight.Symbol into g
-                                  // For direction, we'll trust the most recent insight
-                                  select g.OrderBy(x => x.GeneratedTimeUtc).LastOrDefault())
-                                  .ToList();
-
-            if (activeInsights.Count == 0)
+            _insightCollection.AddRange(insights);
+
+            // Create flatten target for each security that was removed from the universe
+            if (_removedSymbols != null)
             {
-                return targets;
+                var universeDeselectionTargets = _removedSymbols.Select(symbol => new PortfolioTarget(symbol, 0));
+                targets.AddRange(universeDeselectionTargets);
+                _removedSymbols = null;
             }
 
+            // Get insight that haven't expired of each symbol that is still in the universe
+            var activeInsights = _insightCollection.GetActiveInsights(algorithm.UtcTime);
+
+            // Get the last generated active insight for each symbol
+            var lastActiveInsights = from insight in activeInsights
+                                     group insight by insight.Symbol into g
+                                     select g.OrderBy(x => x.GeneratedTimeUtc).Last();
+
             // give equal weighting to each security
-            var count = activeInsights.Count(x => x.Direction != InsightDirection.Flat);
+            var count = lastActiveInsights.Count(x => x.Direction != InsightDirection.Flat);
             var percent = count == 0 ? 0 : 1m / count;
 
-            foreach (var insight in activeInsights)
+            foreach (var insight in lastActiveInsights)
             {
                 targets.Add(PortfolioTarget.Percent(algorithm, insight.Symbol, (int)insight.Direction * percent));
             }
 
-            // add targets to remove invested securities
-            targets.AddRange(from kvp in algorithm.Portfolio
-                             where kvp.Value.Invested
-                             where targets.All(x => kvp.Key != x.Symbol)
-                             select new PortfolioTarget(kvp.Key, 0));
+            // Get expired insights and create flatten targets for each symbol
+            var expiredInsights = _insightCollection.RemoveExpiredInsights(algorithm.UtcTime);
+
+            var expiredTargets = from insight in expiredInsights
+                                 group insight.Symbol by insight.Symbol into g
+                                 where !_insightCollection.HasActiveInsights(g.Key, algorithm.UtcTime)
```

Comment:
Is this check require? Looks like we untrust the return value of `RemoveExpiredInsights`
Also, I was thinking that maybe moving the call to `RemoveExpiredInsights` before `GetActiveInsights` would avoid some double work